### PR TITLE
[Fork tests] Adding individual block numbers and networks

### PR DESCRIPTION
**Pull request by jubeira**
This PR allows:
- Running individual fork tests by specifying a task ´id´ as a parameter in the deployments package.
- Setting the block number and the network inside each individual test.

### CODE REVIEW

File: pkg/deployments/hardhat.config.ts [hunk 1]
```diff
@@ -191,6 +191,7 @@ task('check-action-ids', `Check that contract action-ids correspond the expected
   });
 
 task(TASK_TEST)
+  .addOptionalParam('id', 'Optional task name of the fork test to run.')
   .addOptionalParam('fork', 'Optional network name to be forked block number to fork in case of running fork tests.')
   .addOptionalParam('blockNumber', 'Optional block number to fork in case of running fork tests.', undefined, types.int)
```
Comment (nventuro) [inline pkg/deployments/hardhat.config.ts:196]: You can remove these now

File: pkg/deployments/src/forkTests.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      const forkingNetworkName = Object.keys(hre.config.networks).find((networkName) => networkName === forkNetwork);
+      if (!forkingNetworkName) throw Error(`Could not find a config for network ${forkNetwork} to be forked`);
+
+      const forkingNetworkConfig = hre.config.networks[forkingNetworkName] as HttpNetworkConfig;
+      if (!forkingNetworkConfig.url)
+        throw Error(`Could not find a RPC url in network config for ${forkingNetworkName}`);
+
+      await hre.network.provider.request({
+        method: 'hardhat_reset',
+        params: [{ forking: { jsonRpcUrl: forkingNetworkConfig.url, blockNumber: blockNumber } }],
```
Comment (nventuro) [inline pkg/deployments/src/forkTests.ts:22]:
```suggestion
        params: [{ forking: { jsonRpcUrl: forkingNetworkConfig.url, blockNumber } }],
```

File: pkg/deployments/src/forkTests.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      });
+
+      const config = hre.network.config as HardhatNetworkConfig;
+      config.forking = { enabled: true, blockNumber: blockNumber, url: forkingNetworkConfig.url };
```
Comment (nventuro) [inline pkg/deployments/src/forkTests.ts:26]:
```suggestion
      config.forking = { enabled: true, blockNumber, url: forkingNetworkConfig.url };
```

File: pkg/deployments/src/test.ts [hunk 4]
```diff
@@ -4,28 +4,14 @@ import { RunSuperFunction, HardhatRuntimeEnvironment, HttpNetworkConfig, Hardhat
...
-
-  const config = hre.network.config as HardhatNetworkConfig;
-  config.forking = { enabled: true, blockNumber: args.blockNumber, url: forkingNetworkConfig.url };
-
+  const idFilter = (file: string, id: string | undefined) => {
+    if (id !== undefined) {
+      return file.includes(id);
+    }
+    return true;
+  };
```
Comment (nventuro) [inline pkg/deployments/src/test.ts:13]:
```suggestion
  const idFilter = (file: string, id: string | undefined) => id ? file.includes(id) : true;
```

File: pkg/deployments/src/test.ts [hunk 5]
```diff
@@ -4,28 +4,14 @@ import { RunSuperFunction, HardhatRuntimeEnvironment, HttpNetworkConfig, Hardhat
...
+  args.testFiles = args.testFiles.filter((file: string) => file.endsWith('.fork.ts') && idFilter(file, args.id));
```
Comment (nventuro) [inline pkg/deployments/src/test.ts:14]: We can drop the `fork.ts` filter, and just do

```
if (args.id) {
  args.testFiles = args.testFiles.filter(file => file.includes(id));
}
```

File: pkg/deployments/tasks/20220628-optimism-root-gauge-factory/test/task.fork.ts [hunk 6]
```diff
@@ -30,13 +31,15 @@ describe('OptimismRootGaugeFactory', function () {
...
-  const { OptimismBAL } = task.input();
+  let task: Task;
+  let OptimismBAL: string;
 
   const VEBAL_HOLDER = '0xCB3593C7c0dFe13129Ff2B6add9bA402f76c797e';
   const GOV_MULTISIG = '0x10A19e7eE7d7F8a52822f6817de8ea18204F2e4f';
 
   before('run task', async () => {
+    task = new Task('20220628-optimism-root-gauge-factory', TaskMode.TEST, getForkedNetwork(hre));
+    OptimismBAL = task.input().OptimismBAL;
```
Comment (nventuro) [inline pkg/deployments/tasks/20220628-optimism-root-gauge-factory/test/task.fork.ts:42]:
```suggestion
    { OptimismBAL } = task.input();
```

**Review by nventuro — COMMENTED** [review]
Looks great so far! Do these tests pass with the block numbers you set?

I left a couple of TS style suggestions for reduced verbosity.

**Comment by jubeira** [comment]
Second pass done.
I've adjusted block numbers considering task deployment blocks for them to pass. 

There are a few outstanding issues to review; this is the summary:
- `DoubleEntrypointFixRelayer` - `exits from the sBTC pool` fails. I'm probably not using `StablePoolEncoder.exitBPTInForExactTokensOut` properly (I had to refactor the existing call to the helper because it doesn't exist anymore).
- `TimelockAuthorizer` --> not yet deployed / no output.
- `SNXRecoveryCoordinator` --> same issue as with `DoubleEntrypointFixRelayer`.
- `MerkleRedeem` (from `vita-merkle`): there's no output but the task does have a date; has it been deployed?

**Review by nventuro — APPROVED** [review]

**Pull request merged**